scarb/src/compiler/plugin/proc_macro/compilation.rs

```diff
@@ -150,6 +155,33 @@ pub fn get_crate_archive_basename(package: &Package) -> Result<String> {
     Ok(format!("{}-{}", package_name, package_version))
 }
 
+pub fn unpack_crate(package: &Package, config: &Config) -> Result<()> {
+    let archive_basename = get_crate_archive_basename(package)?;
+    let archive_name = format!("{}.crate", archive_basename);
+
+    let tar = package.target_path(config).into_child("package").open_ro(
```

Comment:
In Cargo, `run_verify` accepts tarball as file lock guard https://github.com/rust-lang/cargo/blob/3c4c0a2b3c9df8f2e97b58358ce1b36ecc81bc5c/src/cargo/ops/cargo_package.rs#L1102

This file lock guard is created here as exclusive lock https://github.com/rust-lang/cargo/blob/3c4c0a2b3c9df8f2e97b58358ce1b36ecc81bc5c/src/cargo/ops/cargo_package.rs#L115

Here, we only have shared lock on tarball, which does not ensure that no two parallel Scarb runs try to extract the same tarball.

Reply:
Addressed in [79d60fe](https://github.com/software-mansion/scarb/pull/1643/commits/79d60fe30be226a8873caa23b99051b6c1c9444f)